File: Data Structures & Algorithms/generate-parentheses/submission-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # brute force DFS
        res = []
        
        def valid(s):
            open = 0
            for c in s:
                open += 1 if c == '(' else -1
                if open < 0:
                    return False
            return not open
        
        def dfs(s):
            if n * 2 == len(s):
                if valid(s):
                    res.append(s)
                return
            
            dfs(s + '(')
            dfs(s + ')')
        
        dfs("")
        return res


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = []
        res = []

        def backtrack(openn, closen):
            if openn == closen == n:
                res.append("".join(stack))
                return
            
            if openn < n:
                stack.append("(")
                backtrack(openn+1, closen)
                stack.pop()
            if closen < openn:
                stack.append(")")
                backtrack(openn, closen+1)
                stack.pop()
        
        backtrack(0, 0)
        return res

File: Data Structures & Algorithms/daily-temperatures/submission-1.cpp
// brute force whatever -> 
// class Solution {
// public:
//     vector<int> dailyTemperatures(vector<int>& temperatures) {
//         int n = temperatures.size();
//         vector<int> res(n); // pre-allocation for the space of vector
//         for (int i = 0; i < n; i++) {
//             int count = 1;
//             int j = i + 1;
//             while (j < n) {
//                 if (temperatures[i] < temperatures[j]) {
//                     break;
//                 }
//                 j++;
//                 count++;
//             }
//             count = (j == n) ? 0 : count; //tenary operation?
//             res[i] = count;
//         }
//         return res;
//     }
// };

// stack
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size(), 0);
        stack<pair<int, int>> stack; //would be a pair (temp, index)

        for (int i = 0; i < temperatures.size(); i++) {
            int t = temperatures[i];
            while (!stack.empty() && t > stack.top().first) {
                pair<int, int> pair = stack.top();
                stack.pop();
                res[pair.second] = i - pair.second;
            }
            stack.push({t, i});
        }
        return res;
    }
};
